h2o-algos/src/main/java/hex/glm/GLM.java

```diff
@@ -2329,48 +2534,50 @@ private void doCompute() {
-          if ((!_parms._HGLM && (_parms._cold_start || (!_parms._lambda_search && _parms._cold_start))) && (i > 0) 
+          if ((!_parms._HGLM && (_parms._cold_start || (!_parms._lambda_search && _parms._cold_start))) && (i > 0)
                   && !_checkPointFirstIter) // default: cold_start for non lambda_search
             coldStart(devHistoryTrain, devHistoryTest);
           Submodel sm = computeSubmodel(submodelCount, _parms._lambda[i], nullDevTrain, nullDevValid);
           if (_checkPointFirstIter)
             _checkPointFirstIter = false;
           double trainDev = sm.devianceTrain; // this is stupid, they are always -1 except for lambda_search=True
           double testDev = sm.devianceValid;
-          devHistoryTest[submodelCount % devHistoryTest.length] = 
+          devHistoryTest[submodelCount % devHistoryTest.length] =
                   (oldDevTest - testDev) / oldDevTest; // only remembers 5
           oldDevTest = testDev;
-          devHistoryTrain[submodelCount % devHistoryTrain.length] = 
+          devHistoryTrain[submodelCount % devHistoryTrain.length] =
                   (oldDevTrain - trainDev) / oldDevTrain;
           oldDevTrain = trainDev;
           if (_parms._lambda[i] < _lmax && Double.isNaN(_lambdaCVEstimate) /** if we have cv lambda estimate we should use it, can not stop before reaching it */) {
             if (_parms._early_stopping && _state._iter >= devHistoryTrain.length) {
               double s = ArrayUtils.maxValue(devHistoryTrain);
               if (s < 1e-4) {
-                Log.info(LogMsg("converged at lambda[" + i + "] = " + _parms._lambda[i]+ "alpha[" + alphaInd + "] = "
-                        + _parms._alpha[alphaInd]+ ", improvement on train = " + s));
+                Log.info(LogMsg("converged at lambda[" + i + "] = " + _parms._lambda[i] + "alpha[" + alphaInd + "] = "
+                        + _parms._alpha[alphaInd] + ", improvement on train = " + s));
                 break; // started overfitting
               }
               if (_validDinfo != null && _parms._nfolds <= 1) { // check for early stopping on test with no xval
                 s = ArrayUtils.maxValue(devHistoryTest);
                 if (s < 0) {
-                  Log.info(LogMsg("converged at lambda[" + i + "] = " + _parms._lambda[i] +  "alpha[" + alphaInd +
-                          "] = " + _parms._alpha[alphaInd]+", improvement on test = " + s));
+                  Log.info(LogMsg("converged at lambda[" + i + "] = " + _parms._lambda[i] + "alpha[" + alphaInd +
+                          "] = " + _parms._alpha[alphaInd] + ", improvement on test = " + s));
                   break; // started overfitting
                 }
               }
             }
           }
 
-          if (_parms._lambda_search && (_parms._score_each_iteration || timeSinceLastScoring() > _scoringInterval)) {
+          if ((_parms._lambda_search || _parms._generate_scoring_history) && (_parms._score_each_iteration || 
```

Comment:
I don't understand the reason why we test `generate_scoring_history` here. 

Shouldn't we score if we have `score_each_iteration=true` (or the other conditions) even if the `generate_scoring_history=false`?

There's probably a good reason but I can't think of any.

Reply:
@tomasfryda The original if statement is this:

if (_parms._lambda_search && (_parms._score_each_iteration || timeSinceLastScoring() > _scoringInterval)) {
            _model._output.setSubmodelIdx(_model._output._best_submodel_idx = submodelCount); // quick and easy way to set submodel parameters
            scoreAndUpdateModel(); // update partial results
          }

This code is run only when lambda_search is on and when some scoring conditions are met.  We agree that it should also take into account when the user sets the _parms._score_iteration_interval.  Hence, the if condition needs to be modified to include this: ((_parms._score_iteration_interval > 0) &&
                  ((_state._iter % _parms._score_iteration_interval) == 0))

The new if condition is now this: 
(_parms._lambda_search && (_parms._score_each_iteration || 
                  timeSinceLastScoring() > _scoringInterval || ((_parms._score_iteration_interval > 0) &&
                  ((_state._iter % _parms._score_iteration_interval) == 0))))

However, this also is a good place to update scoring history not only when lambda_search = True but also when lambda_search=False but generate_scoring_history=True.  Hence, I added in the additional condition 
_parms._generate_scoring_history check as well to force scoring history generation when lambda_search=False but generate_scoring_history=True.  Hence, we arrived at this totally messy if condition: 

((_parms._lambda_search || _parms._generate_scoring_history) && (_parms._score_each_iteration || 
                  timeSinceLastScoring() > _scoringInterval || ((_parms._score_iteration_interval > 0) &&
                  ((_state._iter % _parms._score_iteration_interval) == 0))))